impress-org/givewp
# Enhancement: Improvements to recurrence in the Donation and Subscription models

## Description

Related: https://github.com/impress-org/give-recurring/pull/1133

This PR aims to improve the experience of dealing with recurrence in the Donation and Subscription models. There were two primary issues to be tackled:
1. Donations didn't have a clear "type" to indicate whether it was singular, the start of a subscription, or a renewal. This is a regular question within business logic and had to be determined via a mix of confusing meta.
2. Subscriptions had an "expiration" method that was terribly confusing. It had nothing to do with when the subscription actually ended, but rather when the next renewal would be. Super confusing.
3. Subscriptions and Initial Donations self-referenced. The proper order for creating a subscription should be to make the subscription and then make the donation (which references the subscription). But in our current database structure the subscription and initial donation both reference each other... This creates a chicken and egg situation.
4. Testing Subscriptions was weird, as the factory for the Subscription was odd (due to the self-referencing).

It was a bit of a tangled mess, which is why this is a single PR instead of 4. The hope though, is that the strangeness is isolated to the repository and the models are easy to use, pointing us towards how we _want_ them to work.

```php
$donation = Donation::find(1);
$donation->type->isRenewal(); // checks whether its a renewal or not

$subscription = Subscription::create(['frequency' => 1, 'period' => SubscriptionPeriod::MONTH());
$subscription->bumpRenewalDate(); // sets the next renewal date; just a helper
$subscription->renewsAt; // gets/sets the next renewal date
```

## Affects

The Donation and Subscription models and repositories.

These are used in the latest Google Analytics release, so we'll need to make a PR for that as well.

## Testing Instructions

We'll want to test anywhere these models are used, which I know is a bit spotty. For the subscription renewals, I wasn't sure what we might need to test/update to make sure the renewal dates are properly set. This should work with legacy code as the underlying data structure wasn't changed (still uses the expiration column).

## Pre-review Checklist
